[Team8_pair1] Step4 중간완료

mia6111 opened the pull request:
백엔드에서 반찬, 베스트반찬 카테고리 생성
프론트에서 Ajax통신으로 데이터 받아온후 렌더링 및 이벤트리스너 추가했습니다.
Spring cacheable 은 아직 하지 않았습니다. 코드리뷰 부탁드려요!

mia6111 committed:
Add 베스트반찬 카테고리, 반찬 도메인 추가

- 테스트 URL 남겨져있음

mia6111 committed:
Setup 베스트반찬 반찬 테스트데이터

mia6111 committed:
Add 베스트반찬 프론트 첫화면

- 랜덤으로 1개 카테고리 데이터 불러옴
- 클릭이벤트 미적용
- 콜백으로만 비동기제어

mia6111 committed:
Change 반찬 불러오는 HTML양식 변경

- 첫 로드시 빈 리스트 6개  붙임

urbanscenery committed:
Change 반찬가격 데이터

- '원' 단위 제거
- Origin 가격 및 Real 가격 다르게

urbanscenery committed:
Change origin가격여부

urbanscenery committed:
Merge branch 'yeontae_step4' into pair1_step4

mia6111 committed:
Refactor best.js async, await 적용

best.js async, await 적용

mia6111 committed:
Refactor best.js 함수 쪼개기

best.js 함수 쪼개기

mia6111 committed:
Refactor 클릭 이벤트 처리 및 하드코딩 제거

클릭 이벤트 처리 및 하드코딩 제거

mia6111 committed:
Merge branch 'mh_step4' into pair1_step4

mia6111 committed:
Merge best.js 주석정리

mia6111 committed:
Add 베스트반찬 클릭이벤트리스너

- 클릭시 class on으로 변경
- refacBest.js는 임시파일이라 삭제

mia6111 committed:
Add 클릭이벤트에 Ajax로 데이터로드 및 캐싱

- ajax로 요청시 li 태그에 data-load 프로퍼티 추가
- data-load가 존재시 ajax요청 함.

mia6111 committed:
Refactor Step4 자바스크립트소스

- getIndex, registClickEvent 를 Common.js로 이동
- best.js에서 중복함수 제거

javajigi commented on src/main/java/codesquad/domain/Banchan.java:
```diff
@@ -0,0 +1,27 @@
+    @ManyToOne
+    private BestBanchan parent;
+    private String imgUrl;
+    private String title;
```
title null 허용인가요?
각 column 설정을 좀 더 철저히 하면 좋을 것 같아요.

javajigi reviewed (changes requested):
피드백 하나 남겨요. cache 적용한 후 다시 피드백 할께요.

crongro commented on src/main/resources/static/js/best.js:
```diff
@@ -0,0 +1,97 @@
+}
+
+const templateInit = async () =>{
+    const randomIdx = Math.floor(Math.random()*6);
```
6숫자보다는 dom의 갯수를 확인해서 동적으로 값을 할당하는게 나을 듯.

crongro commented on src/main/resources/static/js/best.js:
```diff
@@ -0,0 +1,97 @@
+
+const templateInit = async () =>{
+    const randomIdx = Math.floor(Math.random()*6);
+    await initBestCategory();
```
맙 to the 소사.
async/await 적용을 이해했나요? 
멋집니다. 동작에 대해서 잘 알아두세요. 현재 가장 많이 쓰는 방법이에요.

crongro commented on src/main/resources/static/js/best.js:
```diff
@@ -0,0 +1,97 @@
+}
+
+const getBestBanchanData = (index) => {
+    index = index || "";
```
defaultparameter 가 무엇인지 찾아서 이런경우 수정해보세요.

crongro commented on src/main/resources/static/js/common.js:
```diff
@@ -15,6 +15,11 @@ const fetchManager = ({ url, method, body, headers, callback }) => {
 }
 
+
+const fetchAsync = ({ url, method, body, headers} ) => (
```
common.js 는 좋습니다.
프로젝트에서도 쓸 수 있도록 계속 범용적인 함수로 발전시켜보세요.

crongro commented on src/main/resources/static/js/best.js:
```diff
@@ -0,0 +1,97 @@
+     $('.tab-btn-box').children[index].setAttribute("data-load", "loaded");
+};
+
+const banchanInit = (data, id) => {
```
함수이름은 늘 동사+명사로 '의도'를 잘 드러내세요.

crongro commented on src/main/resources/static/js/best.js:
```diff
@@ -0,0 +1,97 @@
+}
+
+const appendHtmlFromData = (data, templateFun, parentElement) => {
+    const html = data.reduce( (accum, cur) => {
```
reduce 사용 잘 했음.

crongro reviewed (approved):
fe코드가 부쩍 좋아졌네요.

mia6111 committed:
Refactor 베스트카테고리 및 반찬 도메인 

- 베스트카테고리 및 반찬 도메인 validation, persistence 조건 추가
- Rename bestBanchan -> bestCategory 
-bestBanchanService 추가 및 Controller에서 호출

mia6111 committed:
Add  Cacheable - BestCategory, Banchan, Category

- @EnableCaching 및 @Cacheable 추가
-best.js 함수명 변경 및 BestCategory 의 id로 Banchan 데이터 요청하도록 변경
-JS formatMoney 함수 추가

mia6111 commented:
- 백엔드 캐시 적용 및 도메인 Annotation 추가
- 프론트 기존 버그 수정

코드 리뷰 부탁드립니다 :-)

javajigi commented on src/main/java/codesquad/service/BestBanchanService.java:
```diff
@@ -0,0 +1,33 @@
+    //@Autowired
+    //private BanchanRepository banchanRepository; .findAllByParentId
+//todo cacheable
+    @Cacheable("banchan")
```
Cache 데이터에 변경이 발생할 경우 Cache를 무효화하기 위해 CacheEvict이라는 설정을 사용할 수 있어요.
이와 관련한 학습하고 추후 적용해 보세요.

javajigi reviewed (approved):
Cache 관련 피드백 하나 남겼어요.
